clean up: Provide better URL encoding primitives.

anxking opened the pull request:
Instead of accepting a dictionary of an unspecified type, the near_stream_message_url bundle of functions was refactored to receive parameters of specific types. All the functions that use them were also refactored to adapt. 

- Since the near_message_url functions only need a little information about the message to create the URL, they were refactored to receive each info separately (realm: Realm, message_id: int, display_recipient: DisplayRecipientT, stream_id: Optional[int], topic_name: Optional[str]). 
- This decision was made, instead of accepting a Message object, because of the event-handling functions that use those, as the functions from  zerver/lib/outgoing_webhook.py for example, because they don´t work with Message objects and the change would affect all the events mechanics in Zulip. 

If the use of a Message object is necessary, I would like the feedback, so I could work on the changes to the event-handling functions to make this possible. 

Fixes: https://github.com/zulip/zulip/issues/25021

Self-reviewed
- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

anxking committed:
clean up: Provide better URL encoding primitives.

Instead of accepting a dictionary of an unspecified
type, the near_stream_message_url bundle of functions
were refactored to receive parameters of specific types.
All the fuctions that use them were also refactored to
adapt. Since the functions only need a little information
about the message to create the URL, they were refactored
to receive each info separately (realm, message_id,
display_recipient, stream_id and topic_name). This decision
was made, instead of accepting a Message object, because
of the event-handling functions that use those, since they
don´t work with Message objects and the change would affect
all the events mechanics in Zulip.

Fixes: #25021

timabbott commented on zerver/lib/url_encoding.py:
```diff
@@ -74,15 +89,17 @@ def near_stream_message_url(realm: Realm, message: Dict[str, Any]) -> str:
+def near_pm_message_url(
+    realm: Realm, message_id: int, display_recipient: List[UserDisplayRecipient]
+) -> str:
+    msg_id = str(message_id)
```
We have a profile policy of avoiding abbreviations, and also not using two similar variables like this. I would do `message_id_string = str(message_id)` here.

timabbott replied:
That said, are these variables even necessary? I'd Kinda have expected these integers to just get coerced to strings anyway when joining...

anxking replied:
If I don´t convert them to string, I get a warning from mypy!

timabbott commented on zerver/lib/url_encoding.py:
```diff
@@ -74,15 +89,17 @@ def near_stream_message_url(realm: Realm, message: Dict[str, Any]) -> str:
+    realm: Realm, message_id: int, display_recipient: List[UserDisplayRecipient]
+) -> str:
+    msg_id = str(message_id)
+    str_user_ids = [str(recipient["id"]) for recipient in display_recipient]
```
You could probably call this `user_id_strings = [str(recipient_id["id"] for recipient_user in ...]` for similar reasons to the above

anxking committed:
name changes: Improving variables names.

Implemented name changes suggested on review for
PR: "clean up: Provide better URL encoding primitives".

anxking commented:
@timabbott Thank you for your feedback! I have implemented your suggested changes.
